lib/pbench/server/database/models/users.py

```diff
@@ -101,16 +102,25 @@ def evaluate_email(self, key: str, value: str) -> str:
         valid = validate_email(value)
         return valid.email
 
+    def add_token(self, token: ActiveTokens):
+        """Add the given token to the database, removing any expired tokens
+
+        Args:
+            token : An ActiveTokens object add for this user
+        """
+        try:
+            self.auth_tokens.append(token)
+            Database.db_session.add(token)
+            Database.db_session.commit()
+        except Exception:
+            Database.db_session.rollback()
+            raise
```

Comment:
The docstring appears to be stale (or there is some subtlety that I'm missing...):  I don't see any code in this function which removes expired tokens.

Btw, what do lines 112 and 113 do, respectively?  (I.e., why do we need them both?  Or, why do we add `token` to the session instead of adding `self`?)

Reply:
Yes, the comment is stale, I'll correct.

I'll have to noodle about line 112, I think it is incorrect.

Reply:
No this is correct.  Line 112 updates the auth_tokens "column" of the "row" in the users table while the `.add(token)` at line 113 creates the "row" in the active_tokens table for the given token.

Reply:
Note that SQLAlchemy keeps the `relationship` collection and the foreign key in sync automatically. If we made sure the foreign key was assigned, we wouldn't need to manually append to the relationship. (I don't know if it'll complain or overwrite if you append a child to the parent relationship which already has a different foreign key assigned, but of course it's always best not to do both anyway.)